src/storage/hummock_sdk/src/table_watermark.rs

```diff
@@ -79,23 +52,43 @@ impl TableWatermarksIndex {
     pub fn new(watermark_direction: WatermarkDirection, committed_epoch: HummockEpoch) -> Self {
         Self {
             watermark_direction,
-            index: Default::default(),
+            staging_watermarks: VecDeque::new(),
+            committed_watermarks: None,
             latest_epoch: committed_epoch,
             committed_epoch,
         }
     }
 
-    pub fn index(&self) -> &HashMap<VirtualNode, BTreeMap<HummockEpoch, Bytes>> {
-        &self.index
+    pub fn new_committed(
+        committed_watermarks: Arc<TableWatermarks>,
+        committed_epoch: HummockEpoch,
+    ) -> Self {
+        Self {
+            watermark_direction: committed_watermarks.direction,
+            staging_watermarks: VecDeque::new(),
+            committed_epoch,
+            latest_epoch: committed_epoch,
+            committed_watermarks: Some(committed_watermarks),
+        }
     }
 
     pub fn read_watermark(&self, vnode: VirtualNode, epoch: HummockEpoch) -> Option<Bytes> {
-        self.index.get(&vnode).and_then(|epoch_watermarks| {
-            epoch_watermarks
-                .upper_bound(Included(&epoch))
-                .value()
-                .cloned()
-        })
+        // iterate from new epoch to old epoch
+        for (watermark_epoch, vnode_watermark_list) in self.staging_watermarks.iter().rev().chain(
```

Comment:
> the time to query a middle epoch is 7 times the main branch. This is because the implementation is linearly searching from latest epoch to old epoch, and therefore is't biased for querying the watermark in the latest epoch

This is a note only: I guess we can use binary search to make query from non-latest epoch faster but given that only internal state table contains table watermark and internal state table reads are all latest, we don't need to worry about that right now.